### ISSUE TITLE
Output state diagram

### ISSUE THREAD

**Issue by naude-r** [spring-projects/spring-statemachine#70]
I would be great to be able to generate a state diagram (dot format?) of the configured state machine. This would be useful for documentation purposes as well as seeing a graphical view of the state machine.

**Comment by jvalkeal** [comment]
Yes, I've been thinking about this also. Dot is prop a best bet to start experimenting with any kind of visualization.

**Comment by naude-r** [comment]
@jvalkeal how do you generate the current ascii diagram?

**Comment by jvalkeal** [comment]
@naude-r I'm using http://ditaa.sourceforge.net/. Had to compile it from trunk because current releases seem to be broken with some features.

**Comment by naude-r** [comment]
@jvalkeal Thank you. the samples can print an ascii diagram using the "print" command. it seems to work with:
```
@Qualifier("stateChartModel")
```

how is that processed?

can take a stab at provided Dot output.

**Comment by jvalkeal** [comment]
There's no magic on that stateChartModel. Content of those files are created as bean at https://github.com/spring-projects/spring-statemachine/blob/master/spring-statemachine-samples/src/main/java/demo/CommonConfiguration.java#L46.

Every sample has a statechartmodel.txt file on its classpath. You can then print content of that file and then I manually use ditaa to provide image png to be used in docs.

**Comment by jvmlet** [comment]
:+1:  for this  feature request.
However, state machine [web]designer  (which I assume is on road-map) should cover this as well.
Great  competitor to .NET Workflow  :)

**Comment by mehmetsalgar** [comment]
I know this is not exactly it is what asked here, but I created a blog entry, instead having an Output State Diagram as a result artifact, using an UML Diagram to create a Spring State Machine configuration and making it a primary artifact for the project.

May be it can be useful for you, you can reach the blog from the link below.

https://mehmetsalgar.wordpress.com/2015/12/14/ajax-spring-web-flow-and-spring-state-machine/

**Comment by jvalkeal** [comment]
@mehmetsalgar wow, that is funky. Think I need to take a closer look over the holidays what those components can do. Thx for pointing out your blog :)

**Comment by WorkingDevel** [comment]
I have started to program a little tool that dumps a state machine configuration to an Eclipse MDT UML2 model file (XML; same as used by `UmlStateMachineModelFactory`). The model file can when be imported into [UML Designer](http://www.umldesigner.org/) or [Papyrus](https://eclipse.org/papyrus/). It is in an early stage, but maybe a starting point...

**Comment by jvalkeal** [comment]
@WorkingDevel didn't see your message here until now.

That is interesting work indeed. I've been thinking that as any config(javaconfig,uml,repository config) is first translated to pojo model(StateMachineModel) we could add feature to dump machine into these pojo's. Then you would not need to use reflection(which is always prone to errors if internals are changed).

One question, how pretty graph models these will produce? As a human it's always easy to draw these models so that graph doesn't look messy, but reversing that so that program does it automatically is, I assume, less ideal when graph is complex. Visualisation from a raw data is always hard!

**Comment by wolframite** [comment]
I started using Graphviz with guru.nidi - graphviz-java: But it works only with very simple state machines, no region support: https://gist.github.com/wolframite/78afb9b6891fe2815252d300020731dd

**Comment by jonybuzz** [comment]
> I started using Graphviz with guru.nidi - graphviz-java: But it works only with very simple state machines, no region support: https://gist.github.com/wolframite/78afb9b6891fe2815252d300020731dd

@wolframite  It seems a good alternative while this enhancement is being developed. Unfortunately the gist is not functional, there are some parts missing. Could you complete it please? Thanks a lot

**Comment by wolframite** [comment]
@jonybuzz You're right, I fixed it. Problem was that the visualization part looks different in my project, because I outsourced the whole state machine configuration into a JSON file with some customizations. Please try it again. It "works on my laptop"

**Comment by michael-buerkle** [comment]
It has been a while since the last activity in this issue. Is there a solution to this? I wasn't (yet) able to find anything.

Any advice would be greatly appreciated.

**Comment by agupta1009** [comment]
@jvalkeal i am not getting how the ascii diagram ie statechartmodel.txt is generated and from which part of the code. can ypu please attached the piece of code alog with logic for the same

**Comment by wellini** [comment]
We need it too

**Comment by dmlause** [comment]
I am seeking advice to this as well. I'll check out some of the options mentioned above, but curious if there are plans for anything integrated?

**Comment by jalotra** [comment]
very simple state machine diagramming : https://gist.github.com/jalotra/cb5bf249aa07f7c78a9b0e383b9bae15
features : 
- supports simple (without heirarchy) state machine


![Image](https://github.com/user-attachments/assets/654ca891-0e5f-4fa5-a7bc-500826fc2b15)
